jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -36,23 +43,61 @@
   /** Helper for loading a .class file. */
   private static class ClassFileLoader extends ClassLoader {
 
-    private final Path classFile;
+    private static final Map<Path, Class<?>> cache = new ConcurrentHashMap<>();
 
-    private ClassFileLoader(Path classFile) {
-      this.classFile = classFile;
+    private final Path rootDirectory;
+    private final Path defaultClassFile;
+
+    private ClassFileLoader(Path defaultClassFile, Path rootDirectory) {
+      this.defaultClassFile = defaultClassFile;
+      this.rootDirectory = rootDirectory;
     }
 
     @Nullable
     @Override
     public Class findClass(@Nullable String name) {
       try {
-        byte[] bytes = Files.readAllBytes(classFile);
-        return defineClass(name, bytes, 0, bytes.length);
-      } catch (IOException | ClassFormatError ignored) {
+        // Name is only ever null when we call findClass manually, and not null otherwise. If null,
+        // we should resolve the correct filename.
+        Path file = (name == null ? defaultClassFile : getPathFromClassName(name));
+        if (cache.containsKey(file)) {
+          return cache.get(file);
+        }
+
+        if (!Files.exists(file)) {
+          // TODO: Log search class failure?
+          return null;
+        }
+
+        byte[] bytes = Files.readAllBytes(file);
+        Class<?> definedClass = defineClass(name, bytes, 0, bytes.length);
+        cache.put(file, definedClass);
+        return definedClass;
+
+      } catch (IOException
+          | ClassFormatError
+          | IndexOutOfBoundsException
```

Comment:
Should `IndexOutOfBoundsException` be ignored?

Reply:
According to the javadoc:

```
* @throws  IndexOutOfBoundsException
*          If either <tt>off</tt> or <tt>len</tt> is negative, or if
*          <tt>off+len</tt> is greater than <tt>b.length</tt>.
```

Sounds to me that it's just another indication that it's not a valid class file.

Reply:
Actually, it sounds like it depends on the parameters that we pass in. So this should never actually be thrown.